File: backend/app/schemas/dataset_snapshot.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator

from app.schemas.annotation_export import AnnotationClassMapItem, AnnotationExportSampleQuery
from app.schemas.training_readiness import TrainingReadinessExportFormat


class DatasetSnapshotExportConfig(BaseModel):
    format: TrainingReadinessExportFormat = "manifest"
    include_empty: bool = False
    class_map: list[AnnotationClassMapItem] = Field(default_factory=list)


class DatasetSnapshotCreate(BaseModel):
    name: str | None = Field(default=None, max_length=160)
    sample_query: AnnotationExportSampleQuery = Field(default_factory=AnnotationExportSampleQuery)
    export_config: DatasetSnapshotExportConfig = Field(default_factory=DatasetSnapshotExportConfig)

    @field_validator("name")
    @classmethod
    def normalize_name(cls, value: str | None) -> str | None:
        normalized = value.strip() if value else ""
        return normalized or None

    @field_validator("sample_query")
    @classmethod
    def normalize_sample_ids(cls, value: AnnotationExportSampleQuery) -> AnnotationExportSampleQuery:
        if value.sample_ids is not None:
            value.sample_ids = sorted(set(value.sample_ids))
        return value


class DatasetSnapshotRead(BaseModel):
    id: int
    dataset_id: int
    dataset_revision: int
    name: str | None = None
    content_sha256: str
    sample_count: int
    annotation_count: int
    class_count: int
    sample_query: AnnotationExportSampleQuery
    export_config: DatasetSnapshotExportConfig
    created_at: datetime


class DatasetSnapshotContent(BaseModel):
    schema_version: Literal[1]
    dataset: dict[str, object]
    sample_query: AnnotationExportSampleQuery
    export_config: DatasetSnapshotExportConfig
    class_map: list[AnnotationClassMapItem]
    tag_catalog: list[dict[str, object]]
    annotation_class_catalog: list[dict[str, object]]
    sample_count: int
    annotation_count: int
    samples: list[dict[str, object]]


class DatasetSnapshotDocument(BaseModel):
    snapshot_id: int
    dataset_id: int
    dataset_revision: int
    created_at: datetime
    content_sha256: str
    content: DatasetSnapshotContent


DatasetSnapshotChangeType = Literal[
    "added",
    "removed",
    "file",
    "metadata",
    "tags",
    "split",
    "annotations",
]


class DatasetSnapshotDiffSummary(BaseModel):
    added: int = 0
    removed: int = 0
    file_changed: int = 0
    metadata_changed: int = 0
    tags_changed: int = 0
    split_changed: int = 0
    annotations_changed: int = 0
    changed_samples: int = 0


class DatasetSnapshotDiffItem(BaseModel):
    sample_id: int
    relative_path: str
    change_types: list[DatasetSnapshotChangeType]
    before: dict[str, object] | None = None
    after: dict[str, object] | None = None


class DatasetSnapshotDiffResponse(BaseModel):
    dataset_id: int
    base_snapshot_id: int
    target_snapshot_id: int
    base_revision: int
    target_revision: int
    summary: DatasetSnapshotDiffSummary
    total: int
    page: int
    page_size: int
    items: list[DatasetSnapshotDiffItem]


class DatasetSnapshotTrainingLabels(BaseModel):
    schema_version: Literal[1]
    snapshot_id: int
    dataset_id: int
    dataset_revision: int
    snapshot_content_sha256: str
    format: TrainingReadinessExportFormat
    include_empty: bool
    class_map: list[AnnotationClassMapItem]
    sample_count: int
    label_sha256: str
    labels: list[dict[str, object]]

File: frontend/src/types/datasetSnapshot.ts
import type { AnnotationClassMapItem, AnnotationExportSampleQuery } from "./annotationExport";

export type DatasetSnapshotExportFormat =
  | "labelme"
  | "coco_detection"
  | "coco_segmentation"
  | "yolo_detection"
  | "yolo_segmentation"
  | "voc"
  | "csv"
  | "manifest";

export interface DatasetSnapshotExportConfig {
  format: DatasetSnapshotExportFormat;
  include_empty: boolean;
  class_map: AnnotationClassMapItem[];
}

export interface DatasetSnapshotCreate {
  name?: string | null;
  sample_query: AnnotationExportSampleQuery;
  export_config: DatasetSnapshotExportConfig;
}

export interface DatasetSnapshot {
  id: number;
  dataset_id: number;
  dataset_revision: number;
  name: string | null;
  content_sha256: string;
  sample_count: number;
  annotation_count: number;
  class_count: number;
  sample_query: AnnotationExportSampleQuery;
  export_config: DatasetSnapshotExportConfig;
  created_at: string;
}

export interface DatasetSnapshotDocument {
  snapshot_id: number;
  dataset_id: number;
  dataset_revision: number;
  created_at: string;
  content_sha256: string;
  content: {
    schema_version: 1;
    dataset: Record<string, unknown>;
    sample_query: AnnotationExportSampleQuery;
    export_config: DatasetSnapshotExportConfig;
    class_map: AnnotationClassMapItem[];
    tag_catalog: Array<Record<string, unknown>>;
    annotation_class_catalog: Array<Record<string, unknown>>;
    sample_count: number;
    annotation_count: number;
    samples: Array<Record<string, unknown>>;
  };
}

export type DatasetSnapshotChangeType =
  | "added"
  | "removed"
  | "file"
  | "metadata"
  | "tags"
  | "split"
  | "annotations";

export interface DatasetSnapshotDiffItem {
  sample_id: number;
  relative_path: string;
  change_types: DatasetSnapshotChangeType[];
  before: Record<string, unknown> | null;
  after: Record<string, unknown> | null;
}

export interface DatasetSnapshotDiffResponse {
  dataset_id: number;
  base_snapshot_id: number;
  target_snapshot_id: number;
  base_revision: number;
  target_revision: number;
  summary: {
    added: number;
    removed: number;
    file_changed: number;
    metadata_changed: number;
    tags_changed: number;
    split_changed: number;
    annotations_changed: number;
    changed_samples: number;
  };
  total: number;
  page: number;
  page_size: number;
  items: DatasetSnapshotDiffItem[];
}

export interface DatasetSnapshotTrainingLabels {
  schema_version: 1;
  snapshot_id: number;
  dataset_id: number;
  dataset_revision: number;
  snapshot_content_sha256: string;
  format: DatasetSnapshotExportFormat;
  include_empty: boolean;
  class_map: AnnotationClassMapItem[];
  sample_count: number;
  label_sha256: string;
  labels: Array<Record<string, unknown>>;
}
